docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+
+This kernel implements a blockwise matrix multiplication of two matrices of shape `[M, K] @ [K, N] = [M, N]`, where each output block is computed in parallel over the CUDA grid. This grid is specified as the `grid` argument to the outer `plgpu.kernel`, and parallelizes over the non-contracting dimensions M, N of the matrix multiplication.
+<!-- #endregion -->
+
+<!-- #region id="KSvqVNdy726B" -->
+![matmul](../../_static/pallas/gpu/pipeline_matmul.svg)
+<!-- #endregion -->
+
+<!-- #region id="10ebHCQ571Fn" -->
+
+Within a block, we run a sequential pipeline using `plgpu.emit_pipeline` that reduces over the contracting dimension K of the matrix multiplication. On each iteration of the pipeline, we load one tile from each input matrix, multiply them, and then store the result in an accumulator Ref (`plgpu.ACC`). `plgpu.ACC` is a special type of Ref that lives in registers and holds the intermediate results of WGMMA. Once we have accumulated over the entire contracting dimension, we write out the result to the output Ref.
+
+To perform the actual matrix multiplication, we call `plgpu.wgmma` with the accumulator, LHS, and RHS sides as arguments in order to push the arguments into the TensorCore pipeline. All WGMMA operations are executed in order, so this can be viewed as pushing operations into a queue. Since `wgmma` is an asynchronous instruction, `plgpu.wgmma_wait(N)` is used to wait until there are no more than N `wgmma` operations left in-flight. In this particular implementation we wait for 1 in-flight WGMMA, meaning that the WGMMA we queue on the current iteration will be waited for on the next iteration.
+- `wgmma` wants it's arguments to be in a specific format, defined in the [CUDA documentation](https://docs.nvidia.com/cuda/parallel-thread-execution/#register-fragments-and-shared-memory-matrix-layouts). These are implemented by the `TilingTransform` and `SwizzleTransform` transformations on the input BlockSpecs. Note that in the future transforms will be inferred automatically by Mosaic GPU and these will not need to be manually specified.
+- We use the `delay_release` parameter in conjunction with `plgpu.wgmma_wait(1)` to always allow one `WGMMA` operation to stay in-flight in order to ensure good TensorCore utilization. Without this, we would be flushing the TensorCore pipeline on every iteration of the kernel.
+<!-- #endregion -->
+
+```python id="6Vf5_VA9iCD1"
+def matmul(a, b, tile_m=128, tile_n=128, swizzle=128):
+  dtype = jnp.float16
+  elems_128b = swizzle // jnp.dtype(dtype).itemsize
+  tile_k = elems_128b
+  grid_m = m // tile_m
+  grid_k = k // tile_k
+  grid_n = n // tile_n
+  assert tile_m % elems_128b == 0
+
+  # Note: Transforms will be inferred automatically
+  # by Mosaic GPU in the future.
+  transforms = (
+          plgpu.TilingTransform((8, elems_128b)),
+          plgpu.SwizzleTransform(128),
+      )
+
+  def kernel(a_gmem, b_gmem, o_gmem, o_smem, acc):
+    def kernel_body(_, a_smem, b_smem):
```

Comment:
kernel_body in kernel is a little confusing? perhaps call it `kernel` and `pipeline_step`?

Reply:
Done.